### PR TITLE
Advanced Indexing Support #1

### PR DESCRIPTION
**Pull request by kc611**
This PR refactors the fancy index selection of of NumPy array, which fixes the following cases:

- When an integer index is just before or just after an array index, the case is supported by our current advanced indexing logic, however is blocked at typing level.

```python
import numba
import numpy as np

@numba.njit
def func(x, idx):
    return x[:, 2, idx, :, :]

a = np.random.random_integers(0, 100, (10, 11, 12, 13, 14))
b = np.random.random_integers(0, 10, (4,))

print(np.allclose(func(a, b), func.py_func(a, b)))
# Before PR: Error, but we have the support
# After PR: True
```

- When an integer index is after an array index with slices in between (multiple indexing subspaces), the case is not supported by  our current advanced indexing logic, however is not being blocked at typing level.

```python
import numba
import numpy as np

@numba.njit
def func(x, idx):
    return x[:, :, idx, :, 4]

a = np.random.random_integers(0, 100, (10, 11, 12, 13, 14))
b = np.random.random_integers(0, 10, (4,))

print(np.allclose(func(a, b), func.py_func(a, b)))
# Before PR: False, but runs so it's a bug
# After PR: raises Error correctly
```

This also clearly separates out the various limitations of advanced indexing in Numba so that we can add support/remove the limitation one at a time.

### CODE REVIEW

File: numba/core/typing/arraydecl.py [hunk 1]
```diff
@@ -38,41 +38,65 @@ def get_array_index_type(ary, idx):
...
+            raise NumbaTypeError("Unsupported array index type %s in %s"
                                  % (ty, idx))
         (right_indices if ellipsis_met else left_indices).append(ty)
 
+    if advanced:
+        if array_indices > 1:
+            # Advanced indexing limitation # 2
+            raise NumbaTypeError("Numba doesn't support more than"
+                                 " one non-scalar array index.")
+
```
Comment (guilhermeleobas) [inline numba/core/typing/arraydecl.py:93]: There's a trailing space in this line

File: numba/tests/test_fancy_indexing.py [hunk 2]
```diff
@@ -3,206 +3,163 @@
...
+        # Note: Numba may not return the same array strides and
+        # contiguity as Numpy
+        self.assertEqual(got.shape, expected.shape)
+        self.assertEqual(got.dtype, expected.dtype)
+        np.testing.assert_equal(got, expected)
+
+        # Check a copy was *really* returned by Numba
+        if got.size:
+            got.fill(42)
+            np.testing.assert_equal(arr, orig)
```
Comment (guilhermeleobas) [inline numba/tests/test_fancy_indexing.py:83]: I think you can remove the if condition here.

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Hi @kc611, thanks for working on this. I gave an initial review which covers most flake8 issues.

File: numba/tests/test_fancy_indexing.py [hunk 3]
```diff
@@ -3,206 +3,163 @@
...
-            for i in range(len(tup) + 1):
-                yield tup[:i] + (Ellipsis,) + tup[i:]
-
-    def check_getitem_indices(self, arr, indices):
-        pyfunc = getitem_usecase
-        cfunc = jit(nopython=True)(pyfunc)
+    # (shape or array, indices)
+    # Every case has exactly one, one-dimensional array,
+    # Otherwise it's not fancy indexing
+    indexing_cases = [
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:16]: I'm not sure that this list of cases covers the same range of cases as previously presented with the generated inputs. I gather this patch is mostly about better enumeration of the failure modes in typing, however, I'm concerned that removing the existing (what appears to be more broad) range of test inputs means that lowering/runtime checks are not being exercised so thoroughly.

File: numba/tests/test_fancy_indexing.py [hunk 4]
```diff
@@ -3,206 +3,163 @@
...
+        # Check a copy was *really* returned by Numba
+        if got.size:
+            got.fill(42)
+            np.testing.assert_equal(arr, orig)
+
+    def check_setitem_indices(self, arr_shape, index):
+        def set_item(array, idx, item):
+            array[idx] = item
+
+        arr = np.random.random_integers(0, 10, size=arr_shape)
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:89]: Tests using RNGs should set the RNG seed to ensure reproducibility.

File: numba/core/typing/arraydecl.py [hunk 5]
```diff
@@ -38,41 +38,65 @@ def get_array_index_type(ary, idx):
     if not isinstance(idx, types.BaseTuple):
         idx = [idx]
 
+    in_subspace = False
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/core/typing/arraydecl.py:41]: Yes, that's needed. It belongs to logic that's basically counting the subspaces i.e. 'no. of contiguous groups of indices that are either integers or integer arrays' . `num_subspaces` simply counts the no. of such groups, `is_subspace` tells us if we're currently in a group or not, if we aren't that means a new group has been encountered and `num_subspaces` should be incremented.

Now even if more than one subspace isn't supported, having this logic basically means we'll need little to no change to this particular logic when we actually do add that support.

File: numba/tests/test_fancy_indexing.py [hunk 6]
```diff
@@ -3,206 +3,163 @@
...
-            (np.array(['abc','def','ghi'], dtype='<U3'),
-             np.array('WXYZ', dtype='<U4')),
-            (np.zeros(3, dtype=complex), np.array(2+3j, dtype=complex)),
-        ]
-
-        for x1, v in inps:
-            x2 = x1.copy()
-            pyfunc(x1, 0, v)
-            cfunc(x2, 0, v)
-            self.assertPreciseEqual(x1, x2)
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/tests/test_fancy_indexing.py:200]: We can re-add this one. It doesn't seem to be connected to the permutation logic.

File: numba/tests/test_fancy_indexing.py [hunk 7]
```diff
@@ -3,183 +3,207 @@
...
+        (slice(4, 5), 3,
+         np.array([True, False, True, False, True, False, False]),
+         1),
+        (3, np.array([True, False, True, False, True, False]),
+         slice(None), slice(4)),
+    ]
+
+    rng = np.random.default_rng(1)
+
+    def generate_random_indices(self):
```
Comment (kc611) [inline numba/tests/test_fancy_indexing.py:39]: A few points regarding why I gave a last try to making this function rather than completely switching to old testing framework, There are following issues with the old testing logic:

1. The old framework tests on an array having only one size across dimensions on top of only having 3 dimensions. (its shape is simply 4x4x4).
2. Nearly ~400 indices are being generated as all the possible combination of the slices (which is unnecessary in my opinion) and a single array index which is same across all the indices (which is limited).
3. The testing for Ellipsis object is actually not testing the ellipsis object :-(, we are adding the ellipsis as a fourth index in a three dimensional array, so it's basically being dropped rather than being actually expanded into slices.
4. Adding dimensions to array to fix the above would basically add a huge number of cases in old logic, a few thousand indexing cases which will take much longer to run and is even more unnecessary. 

This function pretty much covers most of the needed logic and cases used previously, reducing the cases to ~80 (20 of each type) with the new ones being added separately.

File: numba/core/typing/arraydecl.py [hunk 8]
```diff
@@ -38,41 +38,65 @@ def get_array_index_type(ary, idx):
...
+            in_subspace = False
         elif isinstance(ty, types.SliceType):
-            pass
+            in_subspace = False
         elif isinstance(ty, types.Integer):
             # Normalize integer index
             ty = types.intp if ty.signed else types.uintp
             # Integer indexing removes the given dimension
             ndim -= 1
             has_integer = True
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/core/typing/arraydecl.py]: Removed it from the code.

File: numba/tests/test_fancy_indexing.py (hunk 2 above)
Comment (kc611) [reply numba/tests/test_fancy_indexing.py]: Removed it.

**Comment by esc** [comment]
Going to use the checklist again for this PR too:


I have checked that:

* [ ] There is no other Pull Request proposing the same change. (If there is
      please highlight this and one of the core developers will pick it up.)
* [ ] The proposed additional NumPy functionality is not deprecated by
      NumPy.
* [ ] Public CI is passing.
* [ ] The patch does not touch files in `numba.core`.
* [ ] The patch is implemented using the high-level extension API only
      (https://numba.readthedocs.io/en/stable/extending/high-level.html).
* [ ] The patch updates the documentation to reflect the new functionality.
      (https://github.com/numba/numba/blob/main/docs/source/reference/numpysupported.rst)
* [ ] The patch is line wrapped to 80 characters.
* [ ] Changes to Python source in the patch are `flake8` compliant.
* [ ] The implementation of the NumPy function reflects that present in the
    NumPy implementation itself (look in the NumPy repository:
      https://github.com/numpy/numpy)...
    * [ ] The implementation has references to the region of the NumPy code
          base that provides the same functionality. The references are URLs
          of the form:
          `https://github.com/numpy/numpy/tree/<SHA>/path/to/file.py:#<lines>`
    * [ ] The code "looks" reasonably similar to the NumPy implementation.
    * [ ] The code uses the same algorithm as the NumPy implementation, if
          it does not it explains why and this justification is reasonable.
    * [ ] The type inference part of the Numba implementation restricts the
          types to those declared in the NumPy docs (or if not present, what
          is accepted in the NumPy implementation itself).
    * [ ] The Numba implementation uses the same variable names for
          arguments as in the NumPy implementation.
    * [ ] The Numba code has commentary present for anything not-obvious by
          inspection.
    * [ ] The Numba code accommodates differences that might occur across
          different NumPy versions in a clear fashion and uses
          `numba.np.numpy_support.numpy_version` to separate the difference
          in implementations.
* [ ] The unit tests for the implementation:
    * [ ] Use the same test inputs as the NumPy unit tests for the same
          functionality. If this is not the case there are comments which
          explain why and this justification is reasonable.
    * [ ] The NumPy unit test inputs have reference URLs associated with
          them to track their origin. The references are URLs of the form:
          `https://github.com/numpy/numpy/tree/<SHA>/path/to/file.py:#<lines>`
    * [ ] Use `unittest` style assertions (not pytest style!).
    * [ ] Check that unsupported/invalid input types are handled as expected
          using the `with self.assertRaises()` pattern.
    * [ ] Explicitly check custom error messages are raised appropriately.
    * [ ] Hit all the branches of the proprosed change (both at type
          inference and run time).
    * [ ] Check that numerically awkward values work as expected e.g.
          `NaN`, `+/-Inf`, `0`, `INT_MAX`, `INT_MIN` etc.
    * [ ] Check that types which can be converted to arrays (assuming NumPy
          accepts this behaviour) work as expected.
    * [ ] Check that passing empty versions of containers e.g. empty tuple
          or 0d array works as expected.
* [ ] The patch works in a simple manual test in my local checkout.
* [ ] I cannot break the code in the patch in my local checkout after
      trying to do so for a minimum of 10 minutes. The error messages seem
      reasonable and there are no "incorrect" results.
* [ ] I think the code introduced in this change is easy to maintain.
* [ ] I understand all the code in this PR.

**Comment by esc** [comment]
After doing some manual `flake8` checking, I would like to suggest the following patch so as not to introduce more `flake8`. violations:

```
diff --git i/numba/core/typing/arraydecl.py w/numba/core/typing/arraydecl.py
index 6099af40ea..24c5b2591c 100644
--- i/numba/core/typing/arraydecl.py
+++ w/numba/core/typing/arraydecl.py
@@ -45,8 +45,9 @@ def get_array_index_type(ary, idx):
     for ty in idx:
         if ty is types.ellipsis:
             if ellipsis_met:
-                raise NumbaTypeError("Only one ellipsis allowed in array indices "
-                                     "(got %s)" % (idx,))
+                raise NumbaTypeError(
+                    "Only one ellipsis allowed in array indices "
+                    "(got %s)" % (idx,))
             ellipsis_met = True
             in_subspace = False
         elif isinstance(ty, types.SliceType):
@@ -70,7 +71,8 @@ def get_array_index_type(ary, idx):
               and isinstance(ty.dtype, (types.Integer, types.Boolean))):
             if ty.ndim > 1:
                 # Advanced indexing limitation # 1
-                raise NumbaTypeError("Multi-dimensional indices are not supported.")
+                raise NumbaTypeError(
+                    "Multi-dimensional indices are not supported.")
             array_indices += 1
             advanced = True
             if not in_subspace:
```

**Comment by esc** [comment]
@kc611 this looks good so far. Regarding the `flake8` cleanliness, I noticed that several of the files that you have touched in this PR are still being excluded with our `.flake8` config, but I did check manually and the files are flake8 compliant. Maybe you could include the following patch, so as to automate flake8 checking for these files in furture:

```
diff --git i/.flake8 w/.flake8
index 0e63904d41..c5b40bb90a 100644
--- i/.flake8
+++ w/.flake8
@@ -98,7 +98,6 @@ exclude =
     numba/tests/test_builtins.py
     numba/tests/test_jitmethod.py
     numba/tests/test_inlining.py
-    numba/tests/test_array_manipulation.py
     numba/tests/test_dummyarray.py
     numba/tests/test_smart_array.py
     numba/tests/test_linalg.py
@@ -110,7 +109,6 @@ exclude =
     numba/tests/test_typeinfer.py
     numba/tests/test_return_values.py
     numba/tests/test_npdatetime.py
-    numba/tests/test_fancy_indexing.py
     numba/tests/support.py
     numba/tests/test_print.py
     numba/tests/test_debug.py
@@ -222,7 +220,6 @@ exclude =
     numba/core/typing/randomdecl.py
     numba/core/typing/setdecl.py
     numba/core/typing/npydecl.py
-    numba/core/typing/arraydecl.py
     numba/core/typing/collections.py
     numba/core/typing/ctypes_utils.py
     numba/core/typing/enumdecl.py
```

**Comment by esc** [comment]
> @kc611 this looks good so far. Regarding the `flake8` cleanliness, I noticed that several of the files that you have touched in this PR are still being excluded with our `.flake8` config, but I did check manually and the files are flake8 compliant. Maybe you could include the following patch, so as to automate flake8 checking for these files in furture:
> 
> ```
> diff --git i/.flake8 w/.flake8
> index 0e63904d41..c5b40bb90a 100644
> --- i/.flake8
> +++ w/.flake8
> @@ -98,7 +98,6 @@ exclude =
>      numba/tests/test_builtins.py
>      numba/tests/test_jitmethod.py
>      numba/tests/test_inlining.py
> -    numba/tests/test_array_manipulation.py
>      numba/tests/test_dummyarray.py
>      numba/tests/test_smart_array.py
>      numba/tests/test_linalg.py
> @@ -110,7 +109,6 @@ exclude =
>      numba/tests/test_typeinfer.py
>      numba/tests/test_return_values.py
>      numba/tests/test_npdatetime.py
> -    numba/tests/test_fancy_indexing.py
>      numba/tests/support.py
>      numba/tests/test_print.py
>      numba/tests/test_debug.py
> @@ -222,7 +220,6 @@ exclude =
>      numba/core/typing/randomdecl.py
>      numba/core/typing/setdecl.py
>      numba/core/typing/npydecl.py
> -    numba/core/typing/arraydecl.py
>      numba/core/typing/collections.py
>      numba/core/typing/ctypes_utils.py
>      numba/core/typing/enumdecl.py
> ```

I took another look at this, and it seems like my vscode may have been misconfigured. Perhaps the only file touched in this PR that is flake clean is:

`numba/tests/test_record_dtype.py`?

File: numba/tests/test_fancy_indexing.py [hunk 9]
```diff
@@ -257,5 +256,194 @@ def check(arr, ind):
...
+        assert expected.base is not orig_base
+
+        # Note: Numba may not return the same array strides and
+        # contiguity as Numpy
+        self.assertEqual(got.shape, expected.shape)
+        self.assertEqual(got.dtype, expected.dtype)
+        np.testing.assert_equal(got, expected)
+
+        # Check a copy was *really* returned by Numba
+        got.fill(42)
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/tests/test_fancy_indexing.py]: Resolved in f1591b9

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @kc611 I think the implementation is much improved by these changes as it categorises the cases which are not supported and the new more involved tests are great! I've reviewed this again and it's mostly just simple things left to resolve, but there's a couple of questions that need some attention too. Thanks again for working on this!

File: numba/tests/test_fancy_indexing.py [hunk 10]
```diff
@@ -257,5 +256,194 @@ def check(arr, ind):
...
+        ]
+        integer_choices = list(np.arange(N))
+
+        indices = []
+
+        # Generate 20 random slice cases
+        for i in range(20):
+            array_idx = self.rng.integers(0, 5, size=15)
+            # Randomly select 4 slices from our list
+            curr_idx = self.rng.choice(slice_choices, size=4).tolist()
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/tests/test_fancy_indexing.py:350]: Ignore this, the RNG obviously carries the state.

File: numba/core/typing/arraydecl.py [hunk 11]
```diff
@@ -34,50 +34,87 @@ def get_array_index_type(ary, idx):
...
+    if advanced:
+        if array_indices > 1:
+            # Advanced indexing limitation # 2
+            msg = "Using more than one non-scalar array index is unsupported."
+            raise NumbaTypeError(msg)
+
+        if num_subspaces > 1:
+            # Advanced indexing limitation # 3
+            msg = ("Using more than one indexing subspace (consecutive group "
+                   "of integer or array indices) is unsupported.")
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/core/typing/arraydecl.py]: Addressed in commit: 03177c1

File: numba/tests/test_fancy_indexing.py [hunk 12]
```diff
@@ -301,5 +301,193 @@ def np_new_axis_setitem(a, idx, item):
...
+
+    def test_unsupported_condition_exceptions(self):
+        # Cases with multi-dimensional indexing array
+        idx = (0, 3, np.array([[1, 2], [2, 3]]))
+        with self.assertRaises(TypingError) as raises:
+            self.check_getitem_indices(self.shape, idx)
+        self.assertIn(
+            'Multi-dimensional indices are not supported.',
+            str(raises.exception)
+        )
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/tests/test_fancy_indexing.py]: Addressed in commit: 03177c1

File: numba/tests/test_fancy_indexing.py [hunk 13]
```diff
@@ -301,5 +301,193 @@ def np_new_axis_setitem(a, idx, item):
...
+        orig = arr.copy()
+        orig_base = arr.base or arr
+
+        expected = get_item(arr, index)
+        got = get_item_numba(arr, index)
+        # Sanity check: In advanced indexing, the result is always a copy.
+        assert expected.base is not orig_base
+
+        # Note: Numba may not return the same array strides and
+        # contiguity as Numpy
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/tests/test_fancy_indexing.py]: No it is an existing issue. (The comment is same as the one in the older tests above.)

File: numba/tests/test_fancy_indexing.py [hunk 14]
```diff
@@ -301,5 +301,193 @@ def np_new_axis_setitem(a, idx, item):
...
+        arr = np.random.randint(0, 11, size=arr_shape)
+        src = arr[index]
+        expected = np.zeros_like(arr)
+        got = np.zeros_like(arr)
+
+        set_item.py_func(expected, index, src)
+        set_item(got, index, src)
+
+        # Note: Numba may not return the same array strides and
+        # contiguity as NumPy
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/tests/test_fancy_indexing.py:432]: Answered above.

File: numba/tests/test_fancy_indexing.py [hunk 15]
```diff
@@ -301,5 +301,193 @@ def np_new_axis_setitem(a, idx, item):
...
+        self.assertEqual(got.dtype, expected.dtype)
+
+        np.testing.assert_equal(got, expected)
+
+    def test_getitem(self):
+        # Cases with a combination of integers + other objects
+        indices = self.indexing_cases
+
+        # Cases with permutations of either integers or objects
+        indices += self.generate_random_indices()
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/tests/test_fancy_indexing.py:443]: Resolved in 70e7fc1

File: numba/tests/test_fancy_indexing.py [hunk 16]
```diff
@@ -257,5 +256,194 @@ def check(arr, ind):
...
+            slice(-1, -N, -2),
+            slice(0, N - 1, None),
+            slice(-1, -N, -2)
+        ]
+        integer_choices = list(np.arange(N))
+
+        indices = []
+
+        # Generate 20 random slice cases
+        for i in range(20):
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/tests/test_fancy_indexing.py]: Seems like there's still some loops that don't use `K` and instead have a hardcoded `range(20)`?

File: numba/tests/test_fancy_indexing.py [hunk 17]
```diff
@@ -301,5 +301,188 @@ def np_new_axis_setitem(a, idx, item):
...
+            self.assertIn(
+                err,
+                str(raises.exception)
+            )
+
+
+# if __name__ == '__main__':
+#     unittest.main()
+x = TestFancyIndexingMultiDim()
+x.test_getitem()
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:488]: This seems to be an unrelated change?

File: numba/tests/test_fancy_indexing.py [hunk 18]
```diff
@@ -301,5 +301,188 @@ def np_new_axis_setitem(a, idx, item):
...
+        for _ in range(K):
+            array_idx = self.rng.integers(0, 5, size=15)
+            # Randomly select 4 slices from our list
+            curr_idx = self.rng.choice(slice_choices, size=4).tolist()
+            # Replace one of the slice with the array index
+            _array_idx = self.rng.choice(4)
+            curr_idx[_array_idx] = array_idx
+            indices.append(tuple(curr_idx))
+        # Generate 20 random integer cases 
+        for i in range(20):
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:358]:
```suggestion
        for _ in range(K):
```

File: numba/tests/test_fancy_indexing.py [hunk 19]
```diff
@@ -301,5 +301,188 @@ def np_new_axis_setitem(a, idx, item):
...
+        K = 20 
+        for _ in range(K):
+            array_idx = self.rng.integers(0, 5, size=15)
+            # Randomly select 4 slices from our list
+            curr_idx = self.rng.choice(slice_choices, size=4).tolist()
+            # Replace one of the slice with the array index
+            _array_idx = self.rng.choice(4)
+            curr_idx[_array_idx] = array_idx
+            indices.append(tuple(curr_idx))
+        # Generate 20 random integer cases
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:357]:
```suggestion
        # Generate K random integer cases 
```

File: numba/tests/test_fancy_indexing.py [hunk 20]
```diff
@@ -301,5 +301,188 @@ def np_new_axis_setitem(a, idx, item):
...
+            array_idx = self.rng.integers(0, 5, size=15)
+            # Randomly select 4 integers from our list
+            curr_idx = self.rng.choice(integer_choices, size=4).tolist()
+            # Replace one of the slice with the array index
+            _array_idx = self.rng.choice(4)
+            curr_idx[_array_idx] = array_idx
+            indices.append(tuple(curr_idx))
+
+        # Generate 20 random ellipsis cases
+        for i in range(20):
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:368]:
```suggestion
        # Generate K random ellipsis cases
        for _ in range(K):
```

File: numba/tests/test_fancy_indexing.py [hunk 21]
```diff
@@ -301,5 +301,188 @@ def np_new_axis_setitem(a, idx, item):
...
+            curr_idx = self.rng.choice(slice_choices, size=4).tolist()
+            # Generate two seperate random indices, replace one with
+            # array and second with Ellipsis
+            _array_idx = self.rng.choice(4, size=2, replace=False)
+            curr_idx[_array_idx[0]] = array_idx
+            curr_idx[_array_idx[1]] = Ellipsis
+            indices.append(tuple(curr_idx))
+
+        # Generate 20 random boolean cases
+        for i in range(20):
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:380]:
```suggestion
        # Generate K random boolean cases
        for _ in range(K):
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @kc611, couple of minor things to resolve else looks good.

File: numba/tests/test_fancy_indexing.py (hunk 17 above)
Comment (kc611) [reply numba/tests/test_fancy_indexing.py]: Resolved in [233b1d1](https://github.com/numba/numba/pull/8238/commits/233b1d1b0d2a40c56083cdbfe8e62c9fede67aaa)

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Comment by stuartarchibald** [comment]
@kc611, OOB @esc recalled from previously that there might be some significant CI runtime increases coming from somewhere in these changes. It's hard to tell from public CI as the CPUs used to run the tests are not consistent. Please could you take a look at the performance of the changed test module before/after this patch and report back? Many thanks. (Am going to mark this as waiting on author again whilst this check is performed).

**Comment by kc611** [comment]
Locally I see about a ~40 seconds increase. For the changed test module `test_fancy_indexing`;

On current `main` before the patch, running `pytest numba/numba/tests/test_fancy_indexing.py` gives ` 10 passed, 5 warnings in 40.04s`
After the patch, running `pytest numba/numba/tests/test_fancy_indexing.py` gives ` 13 passed, 201 warnings in 83.21s (0:01:23) `

This is due to the extra tests we have added for multidimensional testing. (Running only the `TestFancyIndexingMultiDim` gives ` 3 passed, 196 warnings in 44.40s` ).

File: numba/tests/test_fancy_indexing.py [hunk 22]
```diff
@@ -301,5 +301,186 @@ def np_new_axis_setitem(a, idx, item):
...
+
+        # Boolean Arrays + Integers
+        (slice(4, 5), 3,
+         np.array([True, False, True, False, True, False, False]),
+         1),
+        (3, np.array([True, False, True, False, True, False]),
+         slice(None), slice(4)),
+    ]
+
+    rng = np.random.default_rng(1)
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:329]: This needs to be in the function that uses it else the sequence generated depends on the order in which the tests are run? i.e. the test class itself is carrying state.

**Review by stuartarchibald — CHANGES_REQUESTED** [review]
Apologies for having to reopen this, but just had a thought RE testing. Please could https://github.com/numba/numba/pull/8238#discussion_r1129415412 be addressed? Many thanks.

**Comment by stuartarchibald** [comment]
> Locally I see about a ~40 seconds increase. For the changed test module `test_fancy_indexing`;
> 
> On current `main` before the patch, running `pytest numba/numba/tests/test_fancy_indexing.py` gives ` 10 passed, 5 warnings in 40.04s` After the patch, running `pytest numba/numba/tests/test_fancy_indexing.py` gives `13 passed, 201 warnings in 83.21s (0:01:23)`
> 
> This is due to the extra tests we have added for multidimensional testing. (Running only the `TestFancyIndexingMultiDim` gives ` 3 passed, 196 warnings in 44.40s` ).

Thanks for doing this @kc611, seems like it is adding a small amount of additional load to the test suite. Maybe the `K` value for the number of random tests could be reduced? I'm also testing this patch locally to try and see the impact on the code base as a whole.

**Comment by stuartarchibald** [comment]
Running this command: `time ./runtests.py -m 4 --random=0.1`.
* `main` branch runs 1161 tests in 286s. real: 4m49, user: 14m18, sys: 0m19
* this branch runs 1161 tests in 478s. real: 8m1, user: 18m44, sys: 0m25

which suggests there may be an issue, will keep looking.

**Comment by stuartarchibald** [comment]
> Running this command: `time ./runtests.py -m 4 --random=0.1`.
> 
>     * `main` branch runs 1161 tests in 286s. real: 4m49, user: 14m18, sys: 0m19
> 
>     * this branch runs 1161 tests in 478s. real: 8m1, user: 18m44, sys: 0m25
> 
> 
> which suggests there may be an issue, will keep looking.

Turns out that this was a mistake in the basis of comparison due to branches containing a slight variation in the tests present, apologies @kc611. Having retested this independently there was no observable performance impact on existing tests.

**Review by stuartarchibald — APPROVED** [review]
Thanks again for the patch and fixes.
